Performance: optimize Sanitizer for the hot path

NickCraver opened the pull request:
I think overall we can replace the Sanitizer with an appropriately formed Regex but this needs a large test suite (started one here - I didn't see it in the WebJobs solution either...and it doesn't appear we're even using the Sanitizer there anymore, so landed test here).

In the hot path especially, almost every call to Sanitize will be a no-op. Since we _always_ look for an equals sign for any token we replace, we can short circuit based on this fact with a single contains check and avoid the test of the `.indexOf` cost in the method. It's not all the love we can give this path, but a non-trivial win for the quick check.

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

#### Crank comparison
| application                             | dev-sw-1    | sanitizer-1 |          |
| --------------------------------------- | ----------- | ----------- | -------- |
| CPU Usage (%)                           |           7 |           7 |    0.00% |
| Cores usage (%)                         |         201 |         201 |    0.00% |
| Working Set (MB)                        |         364 |         374 |   +2.75% |
| Private Memory (MB)                     |       1,102 |       1,158 |   +5.08% |
| Build Time (ms)                         |      17,373 |      14,049 |  -19.13% |
| Start Time (ms)                         |       4,768 |       4,800 |   +0.67% |
| Published Size (KB)                     |     674,686 |     674,686 |    0.00% |
| .NET Core SDK Version                   |     6.0.101 |     6.0.101 |          |
| Max CPU Usage (%)                       |          99 |          99 |    0.00% |
| Max Working Set (MB)                    |         379 |         392 |   +3.43% |
| Max GC Heap Size (MB)                   |         185 |         155 |  -16.22% |
| Size of committed memory by the GC (MB) |         191 |         205 |   +7.33% |
| Max Number of Gen 0 GCs / sec           |       14.00 |       14.00 |    0.00% |
| Max Number of Gen 1 GCs / sec           |        5.00 |        5.00 |    0.00% |
| Max Number of Gen 2 GCs / sec           |        1.00 |        1.00 |    0.00% |
| Max Time in GC (%)                      |        7.00 |        8.00 |  +14.29% |
| Max Gen 0 Size (B)                      |  38,895,824 |  96,901,928 | +149.13% |
| Max Gen 1 Size (B)                      |  10,841,312 |  13,631,560 |  +25.74% |
| Max Gen 2 Size (B)                      |  28,461,904 |  27,379,096 |   -3.80% |
| Max LOH Size (B)                        |   8,830,720 |   9,976,920 |  +12.98% |
| Max Allocation Rate (B/sec)             | 335,214,208 | 337,751,424 |   +0.76% |
| Max GC Heap Fragmentation               |          70 |          82 |  +16.84% |
| # of Assemblies Loaded                  |         201 |         201 |    0.00% |
| Max Exceptions (#/s)                    |         249 |         264 |   +6.02% |
| Max Lock Contention (#/s)               |         127 |         125 |   -1.57% |
| Max ThreadPool Threads Count            |          30 |          28 |   -6.67% |
| Max ThreadPool Queue Length             |          94 |          95 |   +1.06% |
| Max ThreadPool Items (#/s)              |      36,332 |      36,611 |   +0.77% |
| Max Active Timers                       |          86 |          94 |   +9.30% |
| IL Jitted (B)                           |   1,054,476 |   1,054,216 |   -0.02% |
| Methods Jitted                          |      10,881 |      10,878 |   -0.03% |
| Avg Allocation Rate (B/sec)             | 297,712,948 | 308,984,212 |   +3.79% |
| 90th Allocation Rate (B/sec)            | 329,426,128 | 335,245,672 |   +1.77% |


| load                   | dev-sw-1 | sanitizer-1 |        |
| ---------------------- | -------- | ----------- | ------ |
| CPU Usage (%)          |        1 |           1 |  0.00% |
| Cores usage (%)        |       37 |          37 |  0.00% |
| Working Set (MB)       |       37 |          38 | +2.70% |
| Private Memory (MB)    |      357 |         357 |  0.00% |
| Start Time (ms)        |        0 |           0 |        |
| First Request (ms)     |      340 |         340 |  0.00% |
| Requests/sec           |    8,055 |       8,132 | +0.96% |
| Requests               |  484,085 |     488,590 | +0.93% |
| Mean latency (ms)      |    13.92 |       13.59 | -2.37% |
| Max latency (ms)       |   198.99 |      190.65 | -4.19% |
| Bad responses          |        0 |           0 |        |
| Socket errors          |        0 |           0 |        |
| Read throughput (MB/s) |     1.24 |        1.25 | +0.81% |
| Latency 50th (ms)      |    10.04 |        9.96 | -0.80% |
| Latency 75th (ms)      |    15.90 |       15.57 | -2.08% |
| Latency 90th (ms)      |    28.15 |       27.08 | -3.80% |
| Latency 99th (ms)      |    64.70 |       60.30 | -6.80% |

cc @brettsam I think we should backport this one to 3.x given the easy win.

NickCraver committed:
Performance: optimize Sanitizer for the hot path

I think overall we can replace the Sanitizer with an appropriately formed Regex but this needs a large test suite (started one here - I didn't see it in the WebJobs solution either...and it doesn't appear we're even using the Sanitizer there anymore, so landed test here).

In the hot path especially, almost every call to Sanitize will be a no-op. Since we _always_ look for an equals sign for any token we replace, we can short circuit based on this fact with a single contains check and avoid the test of the indexOf cost in the method. It's not all the love we can give this path, but a non-trivial win for the quick check.

mathewc commented on src/WebJobs.Script/Sanitizer.cs:
```diff
@@ -30,6 +30,14 @@ internal static string Sanitize(string input)
+            // Everything we *might* replace contains an equal, so if we don't have that short circuit out.
+            // This can be likely be more efficient with a Regex, but that's best done with a large test suite and this is
+            // a quick/simple win for the high traffic case.
+            if (!input.Contains("="))
```
If someone were to add a new pattern to CredentialTokens (e.g. "Secret:") this check would bypass that. To be future safe, we should add a meta test ensuring that all CredentialTokens contain this skip token.

E.g. move this check into a ContainsCredentials helper, and put a test in place that ensures ContainsCredentials is true for all CredentialTokens

NickCraver replied:
Good idea - will do!

NickCraver replied:
I went with `MayContainCredentials` (better ideas?) after coding it up here to better reflect the meaning (I think) and test added to ensure we don't regress here.

NickCraver committed:
Sanitizer: additional tests & future protections

Ensuring our performance shortcut doesn't skip an addition we make later - a test will fail if that happens forcing us to realize and adjust accordingly.

NickCraver committed:
Merge remote-tracking branch 'origin/dev' into craver/sanitizer-quick-wins
